Unify focus styles, and pave the way for them to be customizable

jasmussen opened the pull request:
This PR does a number of things, the most important one perhaps just being a ton of code cleanup and simplification. It:

- Unifies all focus styles to a single style — a thicker blue border. 
- It creates a new SCSS variable that defines the stroke thickness of this focus border.
- It retires nearly all the focus mixins we used to use, in favor of just these unified styles. 
- It removes a ton of dead code, and reduces specificity selectors where they aren't needed anymore.
- It changes the default focus stroke-width to 1.5px in width. This works as a shape change in addition to the color change, but it's balanced against the simplified UI. 
- <strike>Additionally, it adds a new focus highlight CSS variable, `	--border-width-focus`, which opens up the door to users customizing these. </strike> — Out for now, can be revisited.

A couple of GIFs:

![focus](https://user-images.githubusercontent.com/1204802/77527406-56441280-6e8c-11ea-859d-2d8b537bf2b9.gif)

![sidebar](https://user-images.githubusercontent.com/1204802/77527421-5cd28a00-6e8c-11ea-9cd3-3b6726561066.gif)

Here, a user customized the CSS variable in the inspector:

![user-customized](https://user-images.githubusercontent.com/1204802/77527433-64922e80-6e8c-11ea-8fe2-d8d55f8abb04.gif)

jasmussen committed:
Add focus thickness variable.

jasmussen committed:
Refine placeholder focus.

jasmussen committed:
Retire a number of unused mixins.

jasmussen committed:
Fix focus style for focused toolbar toggle.

jasmussen committed:
Retire switch mixin, unify focus styles.

jasmussen committed:
Unify input focus styles.

jasmussen committed:
Retire dashed line focus style mixin.

jasmussen committed:
Remove menu item focus mixin.

jasmussen committed:
Remove the block style mixins.

jasmussen committed:
Enable a CSS variable to customize the focus width.

jasmussen committed:
Unify focus of custom select box.

jasmussen commented on packages/base-styles/_mixins.scss:
```diff
@@ -114,124 +114,35 @@
-@mixin block-style__is-active() {
-	color: $white;
-	background: $dark-gray-primary;
+@mixin input-style__focus() {
```
These mixins arguably should also be removed. However they are fairly widely used, and in places that are affected by upstream CSS bleed. That's why I omitted these for now. They can be revisited separately.

jasmussen commented on packages/base-styles/_variables.scss:
```diff
@@ -16,6 +16,12 @@ $editor-line-height: 1.8;
+$border-width-focus: 1.5px;
+$border-width-tab: 4px;
+
+:root {
```
Is there a better place than this SCSS file to output these? I imagine there is, because right now the variable appears duplicated a bunch of times, and it shouldn't be. I would appreciate advice here.

youknowriad replied:
What's the reasoning for using a CSS variable here. It seems inconsistent with the other variables (gris, colors...)

jasmussen replied:
They're both variables, that's all the reasoning there is. And that's why I was asking for advice on what might be a better place. It also gets duplicated when here, so it really isn't the best place — just not sure where to put it.

youknowriad replied:
I think the best way to do it would be something like

```
$border-width-focus: var( --border-width-focus, 1.5px );
```

and just use `$border-width-focus` everywhere without this root rule.

That said, I'd prefer if we avoid introducing these variables in this PR in order to give it more thoughts and also be consistent with the variables we propose.

jasmussen replied:
The reason for doing it the way it's been done currently, is to provide a fallback for IE11.

youknowriad replied:
It can potentially be automated using a Post CSS plugin. Maybe by replacing `var` above with a custom one `config` or something

jasmussen replied:
Good argument for leaving it to a separate PR.

jasmussen committed:
Retire CSS variable for now.

jasmussen commented:
Okay, rebased and removed the CSS variable for now. 

I still think we should create a CSS variable for the focus style, so that it can be easily customized for example in options. But that is perhaps something for a different time.

mapk commented:
I tried this out today. In Chrome the focus states look wonderful. I did notice that the checkboxes and color circles still didn't reflect this new change.

![Screen Shot 2020-03-27 at 8 25 11 AM](https://user-images.githubusercontent.com/617986/77771916-8dd5ca80-7004-11ea-9e0e-c4cc18b64cf3.png)

and

![Screen Shot 2020-03-27 at 8 25 23 AM](https://user-images.githubusercontent.com/617986/77771927-91695180-7004-11ea-801f-2d3798045c1c.png)

But that can always be another PR.

In Firefox however, the focus states seemed a bit awkward. They were even around the element. Some items showed a stronger bottom-right shadow, while others showed a thicker top border.

![firefox](https://user-images.githubusercontent.com/617986/77772018-b4940100-7004-11ea-894e-9e138269fff2.gif)

jasmussen commented:
Thank you Mark. I'm having a hard time reproducing the bottom right shadow issue in Firefox. The GIF you're showing has a thicker bottom border to indicate the active tab, this just happens to be the same color as the focus style. But I'm open to suggestions for how to tweak the details here — it feels important to keep the bottom border thickness even while showing focus.

I did notice a little animation flicker that I'll see if I can address!

jasmussen committed:
Address jumpiness.

jasmussen commented:
Addressed the jumpiness.

Forgot to mention that yes, good call on the color swatches and checkboxes. There are some elements inside the canvas as well, like when a social link _button_ (not the block) has focus.

But as you allude to, I'd like to revisit those separately as this PR is becoming a little big.

mkaz reviewed (approved):
👍 Tested it out and it looks good tested on Edge and FF on Win10.
Similar color swatch comments and social links as you mentioned but other focuc styles look fine
